heed/src/cookbook.rs

```diff
@@ -0,0 +1,450 @@
+//!             None => return Err("invalid log key: cannot extract log level".into()),
+//!         };
+//!
+//!         Ok(LogKey { timestamp, level })
+//!     }
+//! }
+//!
+//! /// Encodes the high part of a timestamp. As it is located
+//! /// at the start of the key it can be used to only return
+//! /// the logs that appeared during a, rather long, period.
+//! pub struct LogAtHalfTimestampCodec;
+//!
+//! impl<'a> BytesEncode<'a> for LogAtHalfTimestampCodec {
+//!     type EItem = u32;
+//!
+//!     /// This method encodes only the prefix of the keys in this particular case, the timestamp.
+//!     fn bytes_encode(half_timestamp: &Self::EItem) -> Result<Cow<[u8]>, BoxedError> {
+//!         Ok(Cow::Owned(half_timestamp.to_be_bytes()[..2].to_vec()))
+//!     }
+//! }
+//!
+//! impl<'a> BytesDecode<'a> for LogAtHalfTimestampCodec {
+//!     type DItem = LogKey;
+//!
+//!     fn bytes_decode(bytes: &'a [u8]) -> Result<Self::DItem, BoxedError> {
+//!         LogKeyCodec::bytes_decode(bytes)
+//!     }
+//! }
+//!
+//! fn main() -> Result<(), Box<dyn Error>> {
+//!     let path = Path::new("target").join("heed.mdb");
+//!
+//!     fs::create_dir_all(&path)?;
+//!
+//!     let env = unsafe {
+//!         EnvOpenOptions::new()
+//!             .map_size(10 * 1024 * 1024) // 10MB
+//!             .max_dbs(3000)
+//!             .open(path)?
+//!     };
+//!
+//!     let mut wtxn = env.write_txn()?;
+//!     let db: Database<LogKeyCodec, Str> = env.create_database(&mut wtxn, None)?;
+//!
+//!     db.put(
+//!         &mut wtxn,
+//!         &LogKey { timestamp: 1608326232, level: Level::Debug },
+//!         "this is a very old log",
+//!     )?;
+//!     db.put(
+//!         &mut wtxn,
+//!         &LogKey { timestamp: 1708326232, level: Level::Debug },
+//!         "fibonacci was executed in 21ms",
+//!     )?;
+//!     db.put(&mut wtxn, &LogKey { timestamp: 1708326242, level: Level::Error }, "fibonacci crashed")?;
+//!     db.put(
+//!         &mut wtxn,
+//!         &LogKey { timestamp: 1708326272, level: Level::Warn },
+//!         "fibonacci is running since 12s",
+//!     )?;
+//!
+//!     // We change the way we want to read our database by changing the key codec.
+//!     // In this example we can prefix search only for the logs between a period of time
+//!     // (the two high bytes of the u32 timestamp).
+//!     let iter = db.remap_key_type::<LogAtHalfTimestampCodec>().prefix_iter(&wtxn, &1708326232)?;
+//!
+//!     // As we filtered the log for a specific
+//!     // period of time we must not see the very old log.
+//!     for result in iter {
+//!         let (LogKey { timestamp: _, level: _ }, content) = result?;
+//!         assert_ne!(content, "this is a very old log");
+//!     }
+//!
+//!     Ok(())
+//! }
+//! ```
+//!
+//! # Change the Environment Size Dynamically
+//!
+//! As you may now, you must specify the maximum size of an LMDB environment when you open it.
```

Comment:
Not sure to understand the « As you may now »; is it because, at some point, we didn’t need to specify this information in heed?
Shouldn’t the cookbook only be focused on the latest version of Heed, though?

```suggestion
//! You must specify the maximum size of an LMDB environment when you open it.
```